src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java

```diff
@@ -117,48 +120,100 @@ public static final Configuration EMPTY_CONFIGURATION() {
+            if ((graphLocation != null && null == graphFormat) || (null == graphLocation && graphFormat != null))
+                throw new IllegalStateException(String.format("The %s and %s must both be specified if either is present",
+                    GREMLIN_TINKERGRAPH_GRAPH_LOCATION, GREMLIN_TINKERGRAPH_GRAPH_FORMAT));
+            initEmptyElementCollections();
+            referenceManager = new NoOpReferenceManager();
+            if (graphLocation != null) loadGraph();
         }
+    }
+
+    private void initEmptyElementCollections() {
+        vertices = new THashMap<>();
+        edges = new THashMap<>();
+        verticesByLabel = new THashMap<>(100);
+        edgesByLabel = new THashMap<>(100);
+    }
+
+    /** implementation note: must start with vertices, because the edges require the vertexRefs to be already present! */
+    private void initElementCollections(OndiskOverflow ondiskOverflow) {
+        final Set<Map.Entry<Long, byte[]>> serializedVertices = ondiskOverflow.allVertices();
+        if (!serializedVertices.isEmpty()) {
+            logger.debug("initializing " + serializedVertices.size() + " vertices from existing storage");
+        }
+        vertices = new THashMap<>(serializedVertices.size());
+        verticesByLabel = new THashMap<>(serializedVertices.size());
+        serializedVertices.iterator().forEachRemaining(entry -> {
+            try {
+                final VertexRef<TinkerVertex> vertexRef = ondiskOverflow.getVertexSerializer().deserializeRef(entry.getValue());
+                vertices.put(vertexRef.id, vertexRef);
+                getElementsByLabel(verticesByLabel, vertexRef.label).add(vertexRef);
+            } catch (IOException e) {
+                throw new RuntimeException("error while initializing vertex from storage: id=" + entry.getKey(), e);
+            }
+        });
 
-        if (graphLocation != null) loadGraph();
+        final Set<Map.Entry<Long, byte[]>> serializedEdges = ondiskOverflow.allEdges();
+        if (!serializedEdges.isEmpty()) {
+            logger.debug("initializing " + serializedEdges.size() + " edges from existing storage");
+        }
+        edges = new THashMap<>(serializedEdges.size());
+        edgesByLabel = new THashMap<>(serializedEdges.size());
+        serializedEdges.iterator().forEachRemaining(entry -> {
+            try {
+                final EdgeRef<TinkerEdge> edgeRef = ondiskOverflow.getEdgeSerializer().deserializeRef(entry.getValue());
+                edges.put(edgeRef.id, edgeRef);
+                getElementsByLabel(edgesByLabel, edgeRef.label).add(edgeRef);
+            } catch (IOException e) {
+                throw new RuntimeException("error while initializing edge from storage: id=" + entry.getKey(), e);
+            }
+        });
+        if (!vertices.isEmpty() || !edges.isEmpty())
+            logger.info("initialized from existing storage: " + toString());
```

Comment:
Maybe this should be `logger.debug`?

Reply:
hmm, I thought about that but would rather keep it on info for now. it's a one-off msg anyway